src/sources/dnstap/dns_message_parser.rs

```diff
@@ -0,0 +1,2000 @@
+        255 => Some(String::from("ANY")),
+        256 => Some(String::from("URI")),
+        257 => Some(String::from("CAA")),
+        258 => Some(String::from("AVC")),
+        259 => Some(String::from("DOA")),
+        260 => Some(String::from("AMTRELAY")),
+        32768 => Some(String::from("TA")),
+        32769 => Some(String::from("DLV")),
+
+        _ => None,
+    }
+}
+
+fn format_bytes_as_hex_string(bytes: &[u8]) -> String {
+    bytes
+        .iter()
+        .map(|e| format!("{:02X}", e))
+        .collect::<Vec<String>>()
+        .join(".")
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{
+        net::{Ipv4Addr, Ipv6Addr},
+        str::FromStr,
+        time::Instant,
+    };
+    use trust_dns_proto::{
+        rr::{
+            dnssec::{
+                rdata::{
+                    dnskey::DNSKEY, ds::DS, nsec::NSEC, nsec3::NSEC3, nsec3param::NSEC3PARAM,
+                    sig::SIG, DNSSECRData,
+                },
+                Algorithm as DNSSEC_Algorithm, DigestType, Nsec3HashAlgorithm,
+            },
+            domain::Name,
+            rdata::{
+                caa::KeyValue,
+                sshfp::{Algorithm, FingerprintType},
+                tlsa::{CertUsage, Matching, Selector},
+            },
+            rdata::{null, CAA, NAPTR, SSHFP, TLSA, TXT},
+        },
+        serialize::binary::Restrict,
+    };
+
+    impl DnsMessageParser {
+        pub fn raw_message_clone(&self) -> Option<&Vec<u8>> {
+            self.raw_message_clone.as_ref()
+        }
+    }
+
+    #[test]
+    fn test_parse_as_query_message() {
+        let raw_dns_message = "szgAAAABAAAAAAAAAmg1B2V4YW1wbGUDY29tAAAGAAE=";
+        if let Ok(raw_qeury_message) = base64::decode(raw_dns_message) {
+            let parse_result = DnsMessageParser::new(raw_qeury_message).parse_as_query_message();
+            assert!(parse_result.is_ok());
+            if let Ok(message) = parse_result {
+                assert_eq!(message.header.qr, 0);
+                assert_eq!(message.question_section.len(), 1);
+                assert_eq!(
+                    message.question_section.first().unwrap().name,
+                    "h5.example.com."
+                );
+                assert_eq!(
+                    &message
+                        .question_section
+                        .first()
+                        .unwrap()
+                        .record_type
+                        .clone()
+                        .unwrap(),
+                    "SOA"
+                );
+            } else {
+                error!("Message is not parsed.");
+            }
+        } else {
+            error!("Invalid base64 encoded data.");
+        }
+    }
+
+    #[test]
+    fn test_parse_as_query_message_with_invalid_data() {
+        if let Err(err) = DnsMessageParser::new(vec![1, 2, 3]).parse_as_query_message() {
+            assert!(err.to_string().contains("unexpected end of input"));
```

Comment:
These assertions make it seem like we should be using structured errors to represent them rather than doing string matching.

Reply:
Yes will change it.

Reply:
Changed.